native_lib/DDGCmdline.c

```diff
@@ -0,0 +1,134 @@
+
+enum DDGSyncCryptoSubkeyIds : int {
+    DDGSyncCryptoPasswordHashSubkey = 1,
+    DDGSyncCryptoStretchedPrimaryKeySubkey,
+};
+
+DDGSyncCryptoResult ddgSyncGenerateAccountKeysWithSecretKey(
+    unsigned char primaryKey[DDGSYNCCRYPTO_SECRET_KEY_SIZE],
+    unsigned char *secretKey,
+    unsigned char protectedSecretKey[DDGSYNCCRYPTO_PROTECTED_SECRET_KEY_SIZE],
+    unsigned char passwordHash[DDGSYNCCRYPTO_HASH_SIZE],
+    const char *userId,
+    const char *password) {
+
+    unsigned char salt[crypto_pwhash_SALTBYTES];
+    unsigned char stretchedPrimaryKey[DDGSYNCCRYPTO_STRETCHED_PRIMARY_KEY_SIZE];
+    unsigned char nonceBytes[crypto_secretbox_NONCEBYTES];
+
+    if (NULL == userId) {
+        return DDGSYNCCRYPTO_INVALID_USERID;
+    }
+
+    if (NULL == password) {
+        return DDGSYNCCRYPTO_INVALID_PASSWORD;
+    }
+
+    memset(salt, 0, crypto_pwhash_SALTBYTES);
+    memcpy(salt, userId, min((size_t)crypto_pwhash_SALTBYTES, strlen(userId)));
+
+    if (0 != crypto_pwhash(primaryKey,
+                           DDGSYNCCRYPTO_PRIMARY_KEY_SIZE,
+                           password,
+                           strlen(password),
+                           salt,
+                           crypto_pwhash_OPSLIMIT_INTERACTIVE,
+                           crypto_pwhash_MEMLIMIT_INTERACTIVE,
+                           crypto_pwhash_ALG_DEFAULT)) {
+
+        return DDGSYNCCRYPTO_CREATE_PRIMARY_KEY_FAILED;
+    }
+
+    if (0 != crypto_kdf_derive_from_key(passwordHash,
+                                        DDGSYNCCRYPTO_HASH_SIZE,
+                                        DDGSyncCryptoPasswordHashSubkey,
+                                        DDGSYNC_PASSWORD_HASH_CONTEXT,
+                                        primaryKey)) {
+
+        return DDGSYNCCRYPTO_CREATE_PASSWORD_HASH_FAILED;
+    }
+
+    if (0 != crypto_kdf_derive_from_key(stretchedPrimaryKey,
+                                        DDGSYNCCRYPTO_STRETCHED_PRIMARY_KEY_SIZE,
+                                        DDGSyncCryptoStretchedPrimaryKeySubkey,
+                                        DDGSYNC_STRETCHED_PRIMARY_KEY_CONTEXT,
+                                        primaryKey)) {
+
+        return DDGSYNCCRYPTO_CREATE_STRETCHED_PRIMARY_KEY_FAILED;
+    }
+
+    randombytes_buf(nonceBytes, crypto_secretbox_NONCEBYTES);
+
+    if (0 != crypto_secretbox_easy(protectedSecretKey,
+                                   secretKey,
+                                   DDGSYNCCRYPTO_SECRET_KEY_SIZE,
+                                   nonceBytes,
+                                   stretchedPrimaryKey)) {
+        return DDGSYNCCRYPTO_CREATE_PROTECTED_SECRET_KEY_FAILED;
+    }
+
+    memcpy(&protectedSecretKey[crypto_secretbox_MACBYTES + DDGSYNCCRYPTO_SECRET_KEY_SIZE], nonceBytes, crypto_secretbox_NONCEBYTES);
+
+    return DDGSYNCCRYPTO_OK;
+}
+
+int main(int argc, char **argv) {
+    if (argc != 4) {
+        printf("Usage: %s secretkey username password\n", argv[0]);
```

Comment:
I'm not so sure about this approach of making this CLI highly tailored to what that test script needs. If anything there changes, you'll need a new CLI. And you're calling `DDGCmdline`, which doesn't describe what it does at all.

Personally, when I talked about this being a "thin wrapper", I imagine it having a few basic commands, more or less matching functions like `crypto_pwhash`, `crypto_kdf_derive_from_key`, and `crypto_secretbox_easy`. For example:
- `ddg_crypto primary_key {user_id} {password}` would output a primary key, from a user and password.
- `ddg_crypto hashed_password {primary_key}` would output a password hash, from a primary key.
- `ddg_crypto protected_encryption_key {primary_key} [{secret_key}]` would output an encryption key, from a primary key and a secret key. A bonus would be to let the secret key be optional, in that case you just generate a new random one.

Such strategy would be more future-proof.

And it would be easier to parse the output, too. Right now, you're just outputing 4 lines with literally cryptic values. How were you planning to parse that?

Reply:
I think it's a good idea, I can circle back to this. For the moment a purpose-built tool is the "low hanging fruit".

For parsing the output I turned it into JSON.